### PR TITLE
Dev/cli11: first working implementation of the C++ xarm-commander

### PR DESCRIPTION
**Pull request by jo-bru**
Simple xarm-commander implementation using [CLI11](https://github.com/CLIUtils/CLI11).
@iomz It would be nice if you could take a look at it next week :)

btw, I'm not sure how pull requests should be structured..
(as I didn't create any issues, I can't link any, but I'll do it next time :))

### CODE REVIEW

**Comment by iomz** [comment]
Nice! I'll have a look!

File: src/xarm-commander.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    app.add_flag("-v, --verbose", print_mode, "Verbose mode."); // TODO: multilevel verbose modes + logger
+
+    app.require_subcommand(1); // set max number of subcommands to 1
+
+    // ===== SUBCOMMANDS =====
+    // Subcommand: motion_enable
+    auto motion_enable = app.add_subcommand("motion_enable", "Motion enable");
+
+    bool enable_option = true;
+    motion_enable->add_option("-e, --enable", enable_option, "Enable or disable");
```
Comment (iomz) [inline src/xarm-commander.cpp]: This should be a flag, not an option.
Fixed in a95ef37

File: src/xarm-commander.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    // ===== FLAGS =====
+    bool print_mode = false;
+    app.add_flag("-v, --verbose", print_mode, "Verbose mode."); // TODO: multilevel verbose modes + logger
+
+    app.require_subcommand(1); // set max number of subcommands to 1
+
+    // ===== SUBCOMMANDS =====
+    // Subcommand: motion_enable
+    auto motion_enable = app.add_subcommand("motion_enable", "Motion enable");
```
Comment (iomz) [inline src/xarm-commander.cpp]: The command/option description usually starts with lowercase.

File: src/xarm-commander.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    int servo_option = 8;
+    motion_enable->add_option("-s, --servo", servo_option, "Choose servo (1-8), 8: enable/disable all servo ");
+
+    motion_enable->callback([&]() {
+        arm = new XArmAPI(port);
+        res = arm->motion_enable(enable_option, servo_option);
+
+        if (print_mode)
+            std::cout << "Motion enable - Response: " << res << "\n";
```
Comment (iomz) [inline src/xarm-commander.cpp]: I added a ternary command construction.

File: src/xarm-commander.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    motion_enable->add_option("-e, --enable", enable_option, "Enable or disable");
+
+    int servo_option = 8;
+    motion_enable->add_option("-s, --servo", servo_option, "Choose servo (1-8), 8: enable/disable all servo ");
+
+    motion_enable->callback([&]() {
+        arm = new XArmAPI(port);
+        res = arm->motion_enable(enable_option, servo_option);
+
+        if (print_mode)
```
Comment (iomz) [inline src/xarm-commander.cpp]: Let's put braces even if it's a single line. This concreteness is adopted by Go, and I believe it makes the code safer while doing no harm.

File: src/xarm-commander.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    int res = 1;
+
+    CLI::App app{"xarm-commander: a command line tool for controlling the xArm7."};
+
+    // print defaults on help
+    app.option_defaults()->always_capture_default();
+
+    // ===== FLAGS =====
+    bool print_mode = false;
+    app.add_flag("-v, --verbose", print_mode, "Verbose mode."); // TODO: multilevel verbose modes + logger
```
Comment (iomz) [inline src/xarm-commander.cpp]: > logger
Yes please, instead of `std:cout` 😄

File: src/xarm-commander.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    bool enable_option = true;
+    motion_enable->add_option("-e, --enable", enable_option, "Enable or disable");
+
+    int servo_option = 8;
+    motion_enable->add_option("-s, --servo", servo_option, "Choose servo (1-8), 8: enable/disable all servo ");
+
+    motion_enable->callback([&]() {
+        arm = new XArmAPI(port);
+        res = arm->motion_enable(enable_option, servo_option);
```
Comment (iomz) [inline src/xarm-commander.cpp]: Are these functions adjacent to the Modbus commands in the spec or are they already "interpreted" in the SDK?
(in terms of the "transparency" we want to achieve)

File: src/xarm-commander.cpp [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            for (int i = 0; i < 6; i++)
+                std::cout << pose[i] << " ";
+            std::cout << "]";
+        }
+    });
+
+    CLI11_PARSE(app, argc, argv);
+
+    std::cout << "\nThanks for using xarm-commander!\n"
+              << std::endl;
```
Comment (iomz) [inline src/xarm-commander.cpp:161]: I know, but these messages are rare to find in prevalent command-line tools 😅

**Review by iomz — APPROVED** [review]
See some commits I pushed.
Generally, you should first create an issue, and create a PR that resolves it.
Let's make the problem smaller and do it in shorter cycles (downside would be that I can be a blocker, but I try not to)

**Comment by iomz** [comment]
I also propose to use a C++ linter (e.g., https://clang.llvm.org/extra/clang-tidy/)
Are you using a specific linter that is not from VS-code's default formatter?

File: src/xarm-commander.cpp (hunk 5 above)
Comment (jo-bru) [reply src/xarm-commander.cpp]: Will be done in branch: [feat/logging](https://github.com/Interactions-HSG/xarm-commander/tree/feat/logging)

File: src/xarm-commander.cpp (hunk 6 above)
Comment (jo-bru) [reply src/xarm-commander.cpp]: seem to be quite adjacent, but needs further investigation.

**Comment by jo-bru** [comment]
> I also propose to use a C++ linter (e.g., https://clang.llvm.org/extra/clang-tidy/)
> Are you using a specific linter that is not from VS-code's default formatter?
Should be further discussed on Friday..

**Pull request merged**